File: src/typescript/src/pipeline_services/prompt/prompt.ts
export class PromptStep {
  public prompt_spec: unknown;

  constructor(prompt_spec: unknown) {
    this.prompt_spec = prompt_spec;
  }

  run(input_value: unknown, adapter: unknown): unknown {
    if (adapter == null) {
      throw new Error("No prompt adapter provided");
    }
    if (typeof adapter !== "function") {
      throw new Error("Prompt adapter must be callable");
    }
    return (adapter as (input_value: unknown, prompt_spec: unknown) => unknown)(input_value, this.prompt_spec);
  }
}

File: src/Python/pipeline_services/prompt/prompt.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Optional


@dataclass
class PromptStep:
    prompt_spec: Any

    def run(self, input_value: Any, adapter: Optional[Callable[[Any, Any], Any]]) -> Any:
        if adapter is None:
            raise ValueError("No prompt adapter provided")
        return adapter(input_value, self.prompt_spec)

